libres/lib/include/ert/enkf/state_map.hpp

```diff
@@ -1,62 +1,47 @@
-#include <ert/util/type_macros.h>
+#include <filesystem>
+#include <mutex>
 #include <vector>
 
 #include <ert/enkf/enkf_types.hpp>
 
-typedef struct state_map_struct state_map_type;
-
-extern "C" state_map_type *state_map_alloc();
-state_map_type *state_map_fread_alloc(const char *filename);
-state_map_type *state_map_fread_alloc_readonly(const char *filename);
-state_map_type *state_map_alloc_copy(const state_map_type *map);
-extern "C" bool state_map_is_readonly(const state_map_type *state_map);
-extern "C" void state_map_free(state_map_type *map);
-extern "C" int state_map_get_size(const state_map_type *map);
-extern "C" realisation_state_enum state_map_iget(const state_map_type *map,
-                                                 int index);
-void state_map_update_undefined(state_map_type *map, int index,
-                                realisation_state_enum new_state);
-void state_map_update_matching(state_map_type *map, int index, int state_mask,
-                               realisation_state_enum new_state);
-extern "C" void state_map_iset(state_map_type *map, int index,
-                               realisation_state_enum state);
-extern "C" bool state_map_equal(const state_map_type *map1,
-                                const state_map_type *map2);
-extern "C" void state_map_fwrite(const state_map_type *map,
-                                 const char *filename);
-extern "C" bool state_map_fread(state_map_type *map, const char *filename);
-std::vector<bool> state_map_select_matching(const state_map_type *map,
-                                            int select_mask, bool select);
-void state_map_set_from_inverted_mask(state_map_type *map,
-                                      const std::vector<bool> &mask,
-                                      realisation_state_enum state);
-void state_map_set_from_mask(state_map_type *map, const std::vector<bool> &mask,
-                             realisation_state_enum state);
-int state_map_count_matching(const state_map_type *state_map, int mask);
-extern "C" bool state_map_legal_transition(realisation_state_enum state1,
-                                           realisation_state_enum state2);
-
-UTIL_IS_INSTANCE_HEADER(state_map);
-
-#endif
+class StateMap {
+    std::vector<int> m_state;
+    mutable std::mutex m_mutex;
+    bool m_read_only{};
+
+    void m_assert_writable() const;
+
+public:
+    StateMap() = default;
+    StateMap(const std::filesystem::path &filename, bool read_only = true);
+    StateMap(const StateMap &other);
+
+    int size() const;
+    bool operator==(const StateMap &) const;
+
+    bool is_readonly() const { return m_read_only; };
+
+    void set(int index, realisation_state_enum new_state);
+    realisation_state_enum get(int index) const;
+    void update_matching(size_t index, int state_mask,
+                         realisation_state_enum new_state);
+    void update_undefined(size_t index, realisation_state_enum new_state) {
```

Comment:
What the methods of StateMap does is not completely obvious to me. Can we have some doxygen documentation?